include/EclipseMonitor/Eth/ReceiptMgr.hpp

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) 2022 Haofan Zheng, Tuan Tran
+// Use of this source code is governed by an MIT-style
+// license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+#pragma once
+
+#include <cstdint>
+#include <tuple>
+
+#include "../Internal/SimpleObj.hpp"
+#include "../Internal/SimpleRlp.hpp"
+
+namespace EclipseMonitor
+{
+namespace Eth
+{
+
+/*
+ * ReceiptMgr parses an rlp-encoded Ethereum receipt, and can determine whether
+ * an event has been emitted by a smart contract
+ *
+ * The structure of an Ethereum receipt can be found in the following link:
+ * https://github.com/ethereum/go-ethereum/blob/master/core/types/receipt.go
+ *
+ * The rlp receipt data is expected to come from the Geth API call
+ * "debug_getRawReceipts". This API call, however, does not encode the entire
+ * receipt but only four fields, namely:
+ * 	- 1. Status
+ * 	- 2. Cumulative gas used
+ * 	- 3. Bloom
+ * 	- 4. Logs
+ *
+ * The logs of each receipt contains the data, if any, relating to events that
+ * were emitted by a contract.
+ *
+ * The structure of a receipt log can be found here:
+ * https://github.com/ethereum/go-ethereum/blob/master/core/types/log.go
+ *
+ * This log structure, too, is also not fully encoded to rlp. The rlp encoding
+ * from the API call only contains the first three fields:
+ *
+ * 	- 1. Address
+ * 	- 2. Topics
+ * 	- 3. Data
+ *
+*/
+
+class ReceiptMgr
+{
+public:
+	static Internal::Obj::Object ParseReceipt(
+		const Internal::Obj::Bytes& rlpBytes
+	)
+	{
+		uint8_t firstByte = rlpBytes[0];
+		Internal::Obj::Bytes innerRlp;
+
+		if (firstByte == 0x01 || firstByte == 0x02)
+		{
+			innerRlp = 	Internal::Obj::Bytes(
+				rlpBytes.data() + 1, rlpBytes.data() + rlpBytes.size());
+		}
+		else
+		{
+			innerRlp = rlpBytes;
+		}
+
+		Internal::Obj::Object receiptObj =
+			Internal::Rlp::ParseRlp(innerRlp.GetVal());
+
+		return receiptObj;
+	}
+
+	static ReceiptMgr FromBytes(
+		const Internal::Obj::Bytes& rlpBytes
+	)
+	{
+		Internal::Obj::Object receiptObj;
+		receiptObj = ParseReceipt(rlpBytes);
+
+		return ReceiptMgr(receiptObj);
+	}
+
+	// param: eventHash
+	// param: address  (contract address)
+	// return tuple <isomitted, logdata>
```

Comment:
Since we are checking if an event is emitted, should this be "emitted" instead of "omitted"?